rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py

```diff
@@ -0,0 +1,293 @@
+            role = self._role_requests[i][1]
+            if role is passer.PasserRole:
+                self.assigned_roles.append(role(robot))
+            elif role is receiver.ReceiverRole:
+                self.assigned_roles.append(role(robot))
+            elif role is striker.StrikerRole:
+                self.assigned_roles.append(role(robot))
+            elif role is ball_move.BallMoveRole:
+                self.assigned_roles.append(role(robot))
+
+    def tick(
+        self, world_state: stp.rc.WorldState
+    ) -> List[Tuple[int, RobotIntent]]:  # (id, intent)
+        """
+        FSM
+         - init: request passer
+         - on filled req: init_roles > tick passer's capture
+         - when passer ready to pass: request receiver
+         - on filled req: init_roles > tick passer pass
+         - on ball passed: tick receiver, release passer role
+         - when receiver done: done
+        """
+
+        role_intents = []
+
+        if self._state == State.INIT:
+            self._role_requests = [
+                (
+                    self._init_cost,
+                    ball_move.BallMoveRole,
+                )
+            ]
+            self._needs_assign = True
+            self._state = State.INIT_TICK
+
+        elif self._state == State.INIT_TICK:
+            if len(self.assigned_roles) == 1:
+                self.init_roles(world_state)
+                self._state = State.BALL_SECURE
+
+        elif self._state == State.BALL_SECURE:
+            # TODO: these lines are a little ugly, any fix?
+            ball_move_role = self.assigned_roles[0]
+            intent = ball_move_role.tick(world_state)
+
+            role_intents = [(ball_move_role.robot.id, intent)]
+
+            if ball_move_role.ready:
+                self._state = State.POSSESSING
+
+        elif self._state == State.POSSESSING:
+            ball_move_role = self.assigned_roles[0]
+            intent = ball_move_role.tick(world_state)
+
+            role_intents = [(ball_move_role.robot.id, intent)]
+
+            # Calculate best shot and find nearest robot to the path and compare to threshold
+            best_shot = striker.StrikerRole()._find_target_point(world_state = world_state, kick_speed = 4.0)
+
+            if self.check_kick(world_state, best_shot, ball_move_role.robot_id):
+                print("SHOOOOOOOOOOOOOTTTTTTING!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+                self._state = State.INIT_SHOOT
+
+            elif ball_move_role.is_done(world_state):
+                self._state = State.INIT_SHOOT
+
+            else:
+                #pass ball to most forward person that can be passed to
+                pass_targets = self.get_pass_order(world_state, ball_move_role.robot_id)
+                for robot in pass_targets:
+                    if self.check_kick(world_state, robot.pose[0:2], ball_move_role.robot_id):
+                        self._state = State.INIT_PASS
```

Comment:
I thought it should pass or shoot probabilistically, this logic is just that if there is not a good shot, then pass. This is fine, but just not what was written in the description of the tactic.